File: framework/src/play-cache/src/main/scala/play/api/cache/SyncCacheApi.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.cache

import javax.inject.Inject

import scala.concurrent.duration.{ Duration, _ }
import scala.concurrent.{ Await, Future }
import scala.reflect.ClassTag

/**
 * A cache API that uses synchronous calls rather than async calls. Useful when you know you have a fast in-memory cache.
 */
trait SyncCacheApi {

  /**
   * Set a value into the cache.
   *
   * @param key Item key.
   * @param value Item value.
   * @param expiration Expiration time.
   */
  def set(key: String, value: Any, expiration: Duration = Duration.Inf): Unit

  /**
   * Remove a value from the cache
   */
  def remove(key: String): Unit

  /**
   * Retrieve a value from the cache, or set it from a default function.
   *
   * @param key Item key.
   * @param expiration expiration period in seconds.
   * @param orElse The default function to invoke if the value was not found in cache.
   */
  def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration = Duration.Inf)(orElse: => A): A

  /**
   * Retrieve a value from the cache for the given type
   *
   * @param key Item key.
   * @return result as Option[T]
   */
  def get[T: ClassTag](key: String): Option[T]
}

/**
 * A SyncCacheApi that wraps an AsyncCacheApi
 */
class DefaultSyncCacheApi @Inject() (val cacheApi: AsyncCacheApi) extends SyncCacheApi {

  protected val awaitTimeout: Duration = 5.seconds

  def set(key: String, value: Any, expiration: Duration): Unit = {
    Await.result(cacheApi.set(key, value, expiration), awaitTimeout)
  }

  def get[T: ClassTag](key: String): Option[T] = {
    Await.result(cacheApi.get(key), awaitTimeout)
  }

  def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration)(orElse: => A): A = {
    Await.result(cacheApi.getOrElseUpdate(key, expiration)(Future.successful(orElse)), awaitTimeout)
  }

  def remove(key: String): Unit = {
    Await.result(cacheApi.remove(key), awaitTimeout)
  }
}

File: framework/src/play-cache/src/main/java/play/cache/SyncCacheApi.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.cache;

import java.util.concurrent.Callable;

/**
 * A synchronous API to access a Cache.
 */
public interface SyncCacheApi {
    /**
     * Retrieves an object by key.
     *
     * @param <T> the type of the stored object
     * @param key the key to look up
     * @return the object or null
     */
    <T> T get(String key);

    /**
     * Retrieve a value from the cache, or set it from a default Callable function.
     *
     * @param <T> the type of the value
     * @param key Item key.
     * @param block block returning value to set if key does not exist
     * @param expiration expiration period in seconds.
     * @return the value
     */
    <T> T getOrElseUpdate(String key, Callable<T> block, int expiration);

    /**
     * Retrieve a value from the cache, or set it from a default Callable function.
     *
     * The value has no expiration.
     *
     * @param <T> the type of the value
     * @param key Item key.
     * @param block block returning value to set if key does not exist
     * @return the value
     */
    <T> T getOrElseUpdate(String key, Callable<T> block);

    /**
     * Sets a value with expiration.
     *
     * @param key Item key.
     * @param value The value to set.
     * @param expiration expiration in seconds
     */
    void set(String key, Object value, int expiration);

    /**
     * Sets a value without expiration.
     *
     * @param key Item key.
     * @param value The value to set.
     */
    void set(String key, Object value);

    /**
     * Removes a value from the cache.
     *
     * @param key The key to remove the value for.
     */
    void remove(String key);
}
